File: vision-mobile/src/engine/regime.ts
import { VIX_HALT } from "../config";
import type { OHLCVBar, Regime } from "./types";
import { ema } from "./indicators";

function trend(bars: OHLCVBar[]): "up" | "down" | "mixed" {
  const close = bars.map((b) => b.close);
  const e50 = ema(close, 50);
  const e200 = ema(close, 200);
  const last = close[close.length - 1];
  const slopeUp = e50[e50.length - 1] > e50[e50.length - 6];
  if (last > e200[e200.length - 1] && slopeUp) return "up";
  if (last < e200[e200.length - 1] && !slopeUp) return "down";
  return "mixed";
}

export function stockRegime(
  spy: OHLCVBar[],
  qqq: OHLCVBar[],
  vixLast: number | null
): Regime {
  const tSpy = trend(spy);
  const tQqq = trend(qqq);
  let mode: Regime["mode"];
  if (vixLast !== null && vixLast > VIX_HALT) {
    mode = "halt";
  } else if (tSpy === "up" && tQqq === "up") {
    mode = "long";
  } else if (tSpy === "down" && tQqq === "down") {
    mode = "short";
  } else {
    mode = "mixed";
  }
  return {
    market: "stocks",
    mode,
    long_allowed: mode === "long" || mode === "mixed",
    short_allowed: mode === "short" || mode === "mixed",
    half_size: mode === "mixed",
    detail: { SPY: tSpy, QQQ: tQqq, VIX: vixLast },
  };
}

export function cryptoRegime(btc: OHLCVBar[]): Regime {
  const close = btc.map((b) => b.close);
  const e50 = ema(close, 50);
  const e200 = ema(close, 200);
  const last = close[close.length - 1];
  let mode: Regime["mode"];
  if (last > e200[e200.length - 1] && last > e50[e50.length - 1]) {
    mode = "long";
  } else if (last < e200[e200.length - 1] && last < e50[e50.length - 1]) {
    mode = "short";
  } else {
    mode = "mixed";
  }
  return {
    market: "crypto",
    mode,
    long_allowed: mode === "long" || mode === "mixed",
    short_allowed: mode === "short" || mode === "mixed",
    half_size: mode === "mixed",
    detail: {
      BTC_vs_EMA200: last > e200[e200.length - 1] ? "above" : "below",
      BTC_vs_EMA50: last > e50[e50.length - 1] ? "above" : "below",
    },
  };
}

File: src/backend/engine/regime.py
"""Filtro di regime (§2 della strategia): il semaforo che decide long/short/stop."""
import pandas as pd

from config import VIX_HALT
from engine.indicators import ema


def _trend(df: pd.DataFrame) -> str:
    """'up' | 'down' | 'mixed' sul daily: prezzo vs EMA200 + pendenza EMA50."""
    close = df["close"]
    e50 = ema(close, 50)
    e200 = ema(close, 200)
    last = close.iloc[-1]
    slope_up = e50.iloc[-1] > e50.iloc[-6]
    if last > e200.iloc[-1] and slope_up:
        return "up"
    if last < e200.iloc[-1] and not slope_up:
        return "down"
    return "mixed"


def stock_regime(spy: pd.DataFrame, qqq: pd.DataFrame, vix_last: float | None) -> dict:
    t_spy, t_qqq = _trend(spy), _trend(qqq)
    if vix_last is not None and vix_last > VIX_HALT:
        mode = "halt"
    elif t_spy == "up" and t_qqq == "up":
        mode = "long"
    elif t_spy == "down" and t_qqq == "down":
        mode = "short"
    else:
        mode = "mixed"
    return {
        "market": "stocks",
        "mode": mode,
        "long_allowed": mode in ("long", "mixed"),
        "short_allowed": mode in ("short", "mixed"),
        "half_size": mode == "mixed",
        "detail": {"SPY": t_spy, "QQQ": t_qqq, "VIX": vix_last},
    }


def crypto_regime(btc: pd.DataFrame) -> dict:
    close = btc["close"]
    e50, e200 = ema(close, 50), ema(close, 200)
    last = close.iloc[-1]
    if last > e200.iloc[-1] and last > e50.iloc[-1]:
        mode = "long"
    elif last < e200.iloc[-1] and last < e50.iloc[-1]:
        mode = "short"
    else:
        mode = "mixed"  # solo BTC/ETH, size dimezzata
    return {
        "market": "crypto",
        "mode": mode,
        "long_allowed": mode in ("long", "mixed"),
        "short_allowed": mode in ("short", "mixed"),
        "half_size": mode == "mixed",
        "detail": {
            "BTC_vs_EMA200": "above" if last > e200.iloc[-1] else "below",
            "BTC_vs_EMA50": "above" if last > e50.iloc[-1] else "below",
        },
    }
